lib/PhpStan/Model/RelationDynamicMethodReturnTypeReflection.php

```diff
@@ -0,0 +1,120 @@
+
+class RelationDynamicMethodReturnTypeReflection implements DynamicMethodReturnTypeExtension
+{
+    public function getClass(): string
+    {
+        $res = Relation::class;
+
+        return $res;
+    }
+
+    public function isMethodSupported(MethodReflection $methodReflection): bool
+    {
+        return in_array($methodReflection->getName(), ['first', 'last', 'find', 'take']);
+    }
+
+    public function getTypeFromMethodCall(MethodReflection $methodReflection, \PhpParser\Node\Expr\MethodCall $methodCall, Scope $scope): Type
+    {
+        $calledOnType = $scope->getType($methodCall->var);
+
+        assert($calledOnType instanceof GenericObjectType);
+        // Here you have access to the generic type
+        $genericTypes = $calledOnType->getTypes();
+
+        assert($genericTypes[0] instanceof TypeWithClassName);
+        $subclass = $genericTypes[0]->getClassName();
+        $args = $methodCall->args;
+
+        $args = array_map(static function ($arg) use ($scope) {
+            assert($arg instanceof Arg);
+            $val = $arg->value;
+
+            return $scope->getType($val);
+        }, $args);
+
+        switch ($methodCall->name) {
+            case 'find':
+                $numArgs = count($args);
+                $single = false;
+                $nullable = false;
+
+                if (1 == $numArgs) {
+                    if (!($args[0] instanceof ConstantArrayType)
+                        || (!$this->isNumericArray($args[0]))) {
+                        $single = true;
+                    }
+                } elseif ($numArgs > 1) {
+                    if (($args[0] instanceof ConstantStringType) && (
+                        'first' === $args[0]->getValue()
+                        || 'last' === $args[0]->getValue()
+                    )) {
+                        $single = true;
+                        $nullable = true;
+                    }
+                }
+
+                if ($single && $nullable) {
+                    return new UnionType([
+                        new ObjectType($subclass),
+                        new NullType()
+                    ]);
+                } elseif ($single) {
+                    return new ObjectType($subclass);
+                }
+
+                return new ArrayType(new IntegerType(), new ObjectType($subclass));
+
+            case 'last':
+            case 'first':
+            case 'take':
+                $numArgs = count($args);
+                if (1 == $numArgs) {
+                    return new ArrayType(
+                        new IntegerType(),
+                        $genericTypes[0]
+                    );
+                }
+
+                return new UnionType([
+                    $genericTypes[0],
+                    new NullType()
+                ]);
+
+            default:
+                throw new \Exception('Unknown method');
+        }
+    }
+
+    protected function isNumericArray(ConstantArrayType $args): bool
```

Comment:
DRY with

lib/PhpStan/Model/ModelDynamicStaticMethodReturnTypeReflection.php::isNumericArray()

Perhaps put it in a super class instead?

Reply:
Yeah, I was being lazy. 982ddee

Note that I still haven't really fully exercise the generic stuff. I'll add some more test files one of these days.